jpa/subsystem/src/main/java/org/jboss/as/jpa/transaction/TransactionUtil.java

```diff
@@ -137,47 +181,50 @@ public SessionSynchronization(EntityManager session, String scopedPuName) {
-                    if (ROOT_LOGGER.isDebugEnabled())
-                        ROOT_LOGGER.debugf(ignored, "ignoring error that occurred while closing EntityManager for %s (", scopedPuName);
-                }
-            }
-            // The TX reference to the entity manager, should be cleared by the TM
-
+            afterCompletionCalled = true;
+            safeCloseEntityManager();
         }
 
         /**
-         * AS7-6586 requires that the container avoid closing the EntityManager while the application
-         * may be using the EntityManager in a different thread.  If the transaction has been rolled
-         * back, will check if the current thread is the Arjuna transaction manager Reaper thread.  It is not
-         * safe to call EntityManager.close from the Reaper thread, so false is returned.
-         *
-         * TODO: switch to depend on JBTM-1556 instead of checking the current thread name.
-         *
-         * @param status of transaction.
-         * @return
+         * After the JTA transaction is ended (Synchronization.afterCompletion has been called) and
+         * the JTA transaction is no longer associated with application thread (application thread called
+         * transaction.rollback/commit/suspend), the entity manager can safely be closed.
          */
-        private boolean safeToClose(int status) {
-            if (Status.STATUS_COMMITTED != status) {
-                return !TxUtils.isTransactionManagerTimeoutThread();
+        private void safeCloseEntityManager() {
+            if ( afterCompletionCalled == true && transactionDisassociatedFromApplication == true) {
+                if (manager != null) {
+                    try {
+                        if (ROOT_LOGGER.isDebugEnabled())
+                            ROOT_LOGGER.debugf("%s: closing entity managersession", getEntityManagerDetails(manager, scopedPuName));
+                        manager.close();
+                    } catch (Exception ignored) {
+                        if (ROOT_LOGGER.isDebugEnabled())
+                            ROOT_LOGGER.debugf(ignored, "ignoring error that occurred while closing EntityManager for %s (", scopedPuName);
+                    }
+                    manager = null;
+                }
             }
+        }
 
-            return true;
+        @Override
+        public synchronized void onEvent(TransactionEvent transactionEvent) {
+            if (transactionEvent.getTypes().contains(EventType.ASSOCIATED)) {
+                // application thread is now associated with JTA transaction
+                transactionDisassociatedFromApplication = false;
+            }
+            else if(transactionEvent.getTypes().contains(EventType.DISASSOCIATING)) {
+                // application thread is no longer associated with JTA transaction
+                transactionDisassociatedFromApplication = true;
+            }
+            safeCloseEntityManager();
```

Comment:
Is this right? We have lots of code which pushes and pops the transaction context suspend/resume. It seems wrong to close the entity manager when this happens